server/src/main/java/org/elasticsearch/indices/TelemetryAwareIndexingMetrics.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexingOperationListener;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.telemetry.TelemetryProvider;
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class TelemetryAwareIndexingMetrics implements IndexingOperationListener {
+
+    private static final Logger logger = LogManager.getLogger(TelemetryAwareIndexingMetrics.class);
+
+    public static final Setting<TimeValue> TELEMETRY_INDEXING_METRICS_TIME_INTERVAL_SETTING = Setting.positiveTimeSetting(
+        "telemetry.indexing.metrics.time.interval",
+        TimeValue.timeValueSeconds(5),
+        Setting.Property.NodeScope
+    );
+
+    private final ThreadPool threadPool;
+    private final TimeValue interval;
+    private final LongCounter indexedDocsTotalMetric;
+    private final AtomicLong indexedDocsCountAggregator = new AtomicLong(0);
```

Comment:
Why are we keeping two counters around? I think it's enough to use `LongCounter` the telemetry module will take care of pushing the metrics asynchronously

Reply:
My worries is in super high concurrency which this listener is exposed to. The instance is created in `IndicesService` and passed down to all indices/shards. If we're **absolutely** sure that telemetry's LongCounter can sustain potential load (entire WRITE pool is writing) and *won't become a bottleneck* on a critical path, then we can significantly simply this - no need in collecting atomic long and publishing thread.